Add Trino 478 release note (#26726)
trinodb/trino: docs/src/main/sphinx/release/release-478.md

# Release 478 (dd Oct 2025)

## General

> * {{breaking}} Move startup logs to `io.trino.bootstrap` package. ({issue}`26721`)

Reviewer: This is an internal implementation detail, so remove it

> * Run Trino on JDK 25.0.0 (build 36). ({issue}`26693`)

Reviewer: This should go in the Docker image section

## Security

## Web UI

## JDBC driver

## Docker image

## CLI

## BigQuery connector

## Blackhole connector

## Cassandra connector

## ClickHouse connector

## Delta Lake connector

> * Fix failure when reading `NULL` map on `json` type. ({issue}`26700`)

Reviewer: Json doesn’t have a concept of maps, so what what’s a null map?

Author: cc: @wendigo @chenjian2664

Reviewer 2: it fixed the issue with reading map columns where the value type is json and the value is null
How about "Fix reading map type when value type is json and value is `null`"

Author: @chenjian2664 Please rephrase to clarify whether it was a failure or a correctness issue. It should start with either "Fix failure when ..." or "Fix incorrect results when …"

Reviewer 2: Thanks.
It is a failure, so should be "Fix failure when reading map type with value type is json and value is `null`"

## Druid connector

## DuckDB connector

## Elasticsearch connector

## Exasol connector

## Faker connector

## Google Sheets connector

## Hive connector

* Add support for reading encrypted Parquet files. ({issue}`24517`, {issue}`9383`)

## Hudi connector

## Iceberg connector

* Fix failure due to column count mismatch when executing `add_files_from_table`
  procedure. ({issue}`26774`)

> * Support using a local staging path for improved performance of writes to sorted

Reviewer: Improve performance when writing sorted tables and the xxx config option is set.

# Release 478 (dd Oct 2025)

## General

* Add support for column lineage in `UNNEST` clause. ({issue}`16946`)
* Add `allowed-retry-policies` configuration property to specify which retry
  policies can be selected by user. ({issue}`26628`)
* Add support for loading plugins from multiple directories. ({issue}`26855`)

> * Allow to drop catalog that failed to load correctly. ({issue}`26918`)

Reviewer 3: Yeah .. also 

to drop an uninitialized catalog

* Improve performance of queries with `ORDER BY`. ({issue}`26725`)
* Improve performance of `MERGE` statement involving `NOT MATCHED` case. ({issue}`26759`)

> * Fix potential incorrect results when reading `row` type. ({issue}`26806`)

Author: No, we use lowercase for type names.

> * Make `metadata.catalogs` table return all catalogs. ({issue}`26918`)

Author: cc: @homar

> * Make `metadata.catalogs` table return all catalogs. ({issue}`26918`)

Reviewer 4: `including uninitialized ones` sounds good, should I make a change ?

* Ensure that queries with and without `EXPLAIN ANALYZE` are planned in the same
  way. ({issue}`26938`)

## Security

* Propagate `queryId` to [Open Policy Agent](/security/opa-access-control)
  authorizer. ({issue}`26851`)

## Web UI

* Add support for filtering queries by `X-Trino-Trace-Token` value in [](/admin/preview-web-interface). ({issue}`26447`)
* Improve rendering performance of large query JSON code in [](/admin/preview-web-interface). ({issue}`26807`)
* Fix query plan default viewport in [](/admin/preview-web-interface). ({issue}`26749`)
* Fix query splits timeline in [](/admin/preview-web-interface) to prevent
  blank screen ({issue}`26920`)

## JDBC driver

* Add support for zstd, brotli, and gzip compressions. ({issue}`26875`)

## Docker image

> * Run Trino on JDK 25.0.0 (build 36). ({issue}`26693`)

Reviewer 5: Why are we treating this as user visible? Because we support attaching external plugins to Trino run from official docker image?

## CLI

> * Add support for zstd, brotli, and gzip compressions. ({issue}`26875`)

Reviewer 6: Drop

## BigQuery connector

## Blackhole connector

## Cassandra connector

## ClickHouse connector

## Delta Lake connector

> * Fix failure when reading `NULL` map on `json` type. ({issue}`26700`)

Reviewer 6: `JSON` type ?

* Deprecate `gcs.use-access-token` in favor of `gcs.auth-type` config property. ({issue}`26681`)

## Druid connector

## DuckDB connector

## Elasticsearch connector

## Exasol connector

## Faker connector

## Google Sheets connector

* Fix potential query failure when `gsheets.delegated-user-email` config property
  is used. ({issue}`26501`)

## Hive connector

* Add support for reading encrypted Parquet files. ({issue}`24517`, {issue}`9383`)

> * Fix ORC writer to ensure that dates and timestamps older than 1582 are read 

Author: The past release note uses actual dates instead of "gregorian" keyword. I replaced 1582 with 1582-10-15. 

cc: @raunaqmorarka @marcinsbd

  correctly by Apache Hive. ({issue}`26507`)
* Deprecate `gcs.use-access-token` in favor of `gcs.auth-type` config property. ({issue}`26681`)

## Hudi connector

* Deprecate `gcs.use-access-token` in favor of `gcs.auth-type` config property. ({issue}`26681`)

## Iceberg connector

* Improve performance when writing sorted tables and the `iceberg.sorted-writing.local-staging-path`
  config option is set. ({issue}`24376`)

> * Produce metrics about execution in the output of `remove_orphan_files` procedure. ({issue}`26661`)

Reviewer 6: Return execution metrics while running `remove_orphan_files` procedure

# Release 478 (dd Oct 2025)

## General

* Add support for column lineage in `UNNEST` clause. ({issue}`16946`)

> * Add `allowed-retry-policies` configuration property to specify which retry

Author: (Ignoring as an optional comment)

  policies can be selected by user. ({issue}`26628`)
* Add support for loading plugins from multiple directories. ({issue}`26855`)
* Add `/v1/integrations/gateway` endpoint for integration with Trino Gateway. ({issue}`26548`)
* Allow to drop catalog that failed to load correctly. ({issue}`26918`)

> * Improve performance of queries with `ORDER BY`. ({issue}`26725`)

Reviewer 3: either "with an ORDER BY clause" or "using an ORDER BY clause"

* Improve performance of `MERGE` statement involving `NOT MATCHED` case. ({issue}`26759`)
* Improve performance of spilling join queries. ({issue}`26076`)
* Fix potential incorrect results when reading `row` type. ({issue}`26806`)
* Make `metadata.catalogs` table return all catalogs. ({issue}`26918`)
* Ensure that queries with and without `EXPLAIN ANALYZE` are planned in the same
  way. ({issue}`26938`)

## Security

> * Propagate `queryId` to [Open Policy Agent](/security/opa-access-control)

Reviewer 3: ```suggestion
* Propagate `queryId` to the [Open Policy Agent](/security/opa-access-control)
```

  authorizer. ({issue}`26851`)

> ## Web UI

Reviewer 3: Also ... not sure by @martint did not want changes for preview UI in the release notes in the past .. however I think we should have them

Reviewer 7: I don't agree with that reasoning at all. We're shipping the preview web UI because we want users to use that and retire the old UI. So it doesn't make sense to hide away the developments happening on that front. Also, the UI being new or not having all the features of the old one doesn't make it "not suitable for production usage". It's easy to toggle back and forth between new and old UI with a button after https://github.com/trinodb/trino/pull/26705 and `web-ui.preview.enabled` is enabled by default.

* Add support for filtering queries by `X-Trino-Trace-Token` value in [](/admin/preview-web-interface). ({issue}`26447`)
* Improve rendering performance of large query JSON in [](/admin/preview-web-interface). ({issue}`26807`)
* Fix rendering of a large query plans in [](/admin/preview-web-interface). ({issue}`26749`)
* Fix rendering of splits timeline for queued queries in [](/admin/preview-web-interface) to prevent
  blank screen ({issue}`26920`)

> ## JDBC driver

Reviewer 3: Remove this and all other empty ones

## Docker image

* Run Trino on JDK 25.0.0 (build 36). ({issue}`26693`)

## CLI

## BigQuery connector

## Blackhole connector

## Cassandra connector

## ClickHouse connector

## Delta Lake connector

* Fix failure when reading `NULL` map on `json` type. ({issue}`26700`)
* Deprecate `gcs.use-access-token` in favor of `gcs.auth-type` config property. ({issue}`26681`)

## Druid connector

## DuckDB connector

## Elasticsearch connector

## Exasol connector

## Faker connector

## Google Sheets connector

> * Fix potential query failure when `gsheets.delegated-user-email` config property

Reviewer 3: ```suggestion
* Fix potential query failure when the `gsheets.delegated-user-email` configuration property
```

  is used. ({issue}`26501`)

## Hive connector

* Add support for reading encrypted Parquet files. ({issue}`24517`, {issue}`9383`)
* Deprecate `gcs.use-access-token` in favor of `gcs.auth-type` config property. ({issue}`26681`)
* Improve performance of queries using complex predicates on `$path` column. ({issue}`27000`)
* Fix ORC writer to ensure that dates and timestamps older than `1582-10-15` are
  read correctly by Apache Hive. ({issue}`26507`)

## Hudi connector

* Deprecate `gcs.use-access-token` in favor of `gcs.auth-type` config property. ({issue}`26681`)

## Iceberg connector

* Improve performance when writing sorted tables and the `iceberg.sorted-writing.local-staging-path`

>   config option is set. ({issue}`24376`)

Reviewer 3: ```suggestion
 is set. ({issue}`24376`)
```

# Release 478 (22 Oct 2025)

## General

* Add support for column lineage in `UNNEST` clauses. ({issue}`16946`)
* Add `retry-policy.allowed` configuration property to specify which retry

>   policies can be selected by the user. ({issue}`26628`)

Reviewer 8: ```suggestion
* Add `retry-policy.allowed` configuration property to specify which query retry
  policies can be selected by the user. ({issue}`26628`)
```

* Add support for loading plugins from multiple directories. ({issue}`26855`)

> * Add the `/v1/integrations/gateway` endpoint for integration with Trino Gateway. ({issue}`26548`)

Reviewer 9: I agree with you that it's undesirable to have an endpoint for each service and it should be in a reusable manner

IMO, the coupling between Trino Gateway and Trino is inevitable. If Trino changes, Trino Gateway **will** need to change. For example, in [this PR](https://github.com/trinodb/trino-gateway/pull/774) because Trino 477 removed some of the jmx stats, which caused Trino Gateway to fail, we had to pin Trino version to 476 and look for workarounds.

For this endpoint specifically, yes the data it returns can be useful for other software as well and should be gateway-agnostic. But I think the point is that **each software should deserve its own dedicated endpoint so that they are not affected by unintentional changes**. Other software owners are welcome, of course, to use this endpoint, but it is their responsibility to make sure their software doesn't break at **all times**, even after we roll out new changes to this endpoint. It's another way of saying there's no SLA on this endpoint if they use it, whereas if we create a generic endpoint, there's implied SLA

At the end of the day, if other query routing logic software needs similar data to be returned, they will likely end up with similar code in their endpoint. Yes this results in duplicate code. But we can look into how to extract common logic to a helper function. At that point, maybe we will find some discrepancies in these seemingly-similar looking endpoints.

> * Add the `/v1/integrations/gateway` endpoint for integration with Trino Gateway. ({issue}`26548`)

Reviewer 8: > But I think the point is that **each software should deserve its own dedicated endpoint so that they are not affected by unintentional changes**.

I appreciate the openness but I doubt that would be want we would actually want to maintain in the project.

BTW trino already exports tons of metrics in standard format (that's why jmx became obsolete), so i don't quite yet understand why we need dedicated endpoints just for selected metrics useful for selected use case

* Allow dropping an uninitialized catalog that failed to load. ({issue}`26918`)
* Improve performance of queries with an `ORDER BY` clause. ({issue}`26725`)
* Improve performance of `MERGE` statements involving a `NOT MATCHED` case. ({issue}`26759`)
* Improve performance of queries with joins which spill to disk. ({issue}`26076`)

> * Fix potential incorrect results when reading `row` type. ({issue}`26806`)

Reviewer 8: Do we put Fix-es first?

* Return all catalogs, including uninitialized ones, for queries from `metadata.catalogs`. ({issue}`26918`)

> * Ensure that queries with and without `EXPLAIN ANALYZE` are planned identically. ({issue}`26938`)

Reviewer 8: ❤️ 

this is a bug fix, so it's best to start with "Fix"

```suggestion
* Fix `EXPLAIN ANALYZE` planning so that it executes with the same plan as would be used to execute the query being analyzed. ({issue}`26938`)
```

> * In row pattern matching, restrict logical navigations to current match in running semantics. ({issue}`26981`)

Reviewer 5: The nomenclature here is technical. Does it match what we us in docs?

# Release 478 (22 Oct 2025)

## General

* Add support for column lineage in `UNNEST` clauses. ({issue}`16946`)
* Add `retry-policy.allowed` configuration property to specify which retry
  policies can be selected by user. ({issue}`26628`)
* Add support for loading plugins from multiple directories. ({issue}`26855`)
* Add the `/v1/integrations/gateway` endpoint for integration with Trino Gateway. ({issue}`26548`)

> * Allow to drop an uninitialized catalog that failed to load. ({issue}`26918`)

Reviewer 10: ```suggestion
* Allow dropping an uninitialized catalog that failed to load. ({issue}`26918`)
```

* Improve performance of queries with an `ORDER BY` clause. ({issue}`26725`)
* Improve performance of `MERGE` statements involving a `NOT MATCHED` case. ({issue}`26759`)

> * Improve performance of spilling join queries. ({issue}`26076`)

Reviewer 10: Should we say something more explicit like "queries with joins which spill to disk"?

* Deprecate the `gcs.use-access-token` configuration property. Use `gcs.auth-type` instead. ({issue}`26681`)

## Google Sheets connector

* Fix potential query failure when the `gsheets.delegated-user-email` configuration property
  is used. ({issue}`26501`)

## Hive connector

* Add support for reading encrypted Parquet files. ({issue}`24517`, {issue}`9383`)
* Deprecate the `gcs.use-access-token` configuration property. Use `gcs.auth-type` instead. ({issue}`26681`)
* Improve performance of queries using complex predicates on `$path` column. ({issue}`27000`)
* Prevent writing invalid dates and timestamps before `1582-10-15` by the ORC writer. ({issue}`26507`)

## Hudi connector

* Deprecate the `gcs.use-access-token` configuration property. Use `gcs.auth-type` instead. ({issue}`26681`)

## Iceberg connector

* Improve performance when writing sorted tables and `iceberg.sorted-writing.local-staging-path`
 is set. ({issue}`24376`)
* Return execution metrics while running the `remove_orphan_files` procedure. ({issue}`26661`)
* Deprecate the `gcs.use-access-token` configuration property. Use `gcs.auth-type` instead. ({issue}`26681`)

> * Collect NDV stats on all columns when replacing tables. ({issue}`26983`)

Author: NDV isn't specific to Iceberg. I rephrased as "distinct values count" to avoid an abbreviation.
